added AsyncIOOSCUDPServer

jsheedy opened the pull request:
This change adds an AsyncIOOSCUDPServer.  Its intended use is as an alternative to the existing threading/forking servers which uses the asyncio module.

The logic of _UDPHandler.handle was moved into a _handle function so the different datagram mechanisms of the socketserver servers and the asyncio server could use the same handling code.

The existing *OSCUDPServer classes all shared the same methods, so they were pushed into a superclass, OSCUDPServer.  There were no logical changes to them other than that refactor.

The new class can be used like this:

loop = asyncio.get_event_loop()
server = AsyncIOOSCUDPServer(server_address, dispatcher, loop)
server.serve()
loop.run_forever()

attwad commented:
That's awesome! thanks for doing that :)
I'm fine with bumping the minimum requirements to 3.4 to get the asyncio module, can you update this file and remove the 3.3 version from it please? This way the CI should succeed.
https://github.com/attwad/python-osc/blob/master/.travis.yml

attwad commented on pythonosc/osc_server.py:
```diff
@@ -18,8 +18,18 @@
 http://docs.python.org/library/socketserver.html
+
+
+Alternatively, the AsyncIOOSCUDPServer server can be integrated with an asyncio event loop:
```
I'm trying to conform to the google style guide for python code, can you break at 80 chars please? thanks

attwad commented on pythonosc/osc_server.py:
```diff
@@ -138,15 +143,36 @@ class ForkingOSCUDPServer(
+  :param dispatcher: a pythonosc.dispatcher.Dispatcher
+  :param loop: an asyncio event loop
+
+  Each UDP message is handled by _handle, the same method as in the OSCUDPServer family of blocking, threading, and forking servers
```
ditto, 80 chars please

attwad commented on pythonosc/osc_server.py:
```diff
@@ -138,15 +143,36 @@ class ForkingOSCUDPServer(
+  def serve(self):
+    """creates a datagram endpoint and registers it with our event loop"""
+    listen = self._loop.create_datagram_endpoint(lambda: self.OSCProtocolFactory(self.dispatcher), local_addr=self._server_address)
+    transport, protocol = self._loop.run_until_complete(listen)
```
you're not using transport and protocol here, might as well just ignore the returned values

attwad commented on pythonosc/osc_server.py:
```diff
@@ -138,15 +143,36 @@ class ForkingOSCUDPServer(
+
+  def serve(self):
+    """creates a datagram endpoint and registers it with our event loop"""
+    listen = self._loop.create_datagram_endpoint(lambda: self.OSCProtocolFactory(self.dispatcher), local_addr=self._server_address)
```
move the lambda on the previous line with 4 spaces indent to fit in 80 chars please

attwad commented on pythonosc/osc_server.py:
```diff
@@ -138,15 +143,36 @@ class ForkingOSCUDPServer(
-  def verify_request(self, request, client_address):
-    """Returns true if the data looks like a valid OSC UDP datagram."""
-    return _is_valid_request(request)
+  class OSCProtocolFactory(asyncio.DatagramProtocol):
```
no need for this class to be publicly accessible, prefix with _ please
class _OSCProtocolFactory

attwad commented on pythonosc/osc_server.py:
```diff
@@ -138,15 +143,36 @@ class ForkingOSCUDPServer(
+    def __init__(self, dispatcher):
+      self.dispatcher = dispatcher
+
+    def datagram_received(self, data, addr):
```
Is this called by named parameters? if not please change addr to unused_addr to make it explicit that you chose to ignore this parameter

attwad commented on pythonosc/osc_server.py:
```diff
@@ -138,15 +143,36 @@ class ForkingOSCUDPServer(
+class AsyncIOOSCUDPServer():
+  """Asyncio version of the OSC UDP Server.
+
+  :param server_address: tuple of (IP address to bind to, port)
```
can you move those params to the constructor documentation please?

attwad commented on pythonosc/osc_server.py:
```diff
@@ -29,6 +39,42 @@
 from pythonosc import osc_packet
 
 
+def _handle(data, dispatcher):
```
We need a better name for this method now that it is shared between different kind of classes, how about _call_handlers_for_packet ?

attwad commented:
Alright I think that's all for my comments, please address and I'll merge the pull request and release a new version on pypi with it :)
Sorry for the many nitpicks, I'll try and find some time to put up a CONTRIBUTING file where I can point to the style guide and other things.
Thanks again!

jsheedy commented:
Thanks for all the feedback, I appreciate it! I will commit fixes for each of the issues.

jsheedy commented:
I believe I addressed all the comments in the last commit, let me know what you think.  Sorry, I neglected to consider 3.3 when I wrote this.  We could support 3.3 by having a dependency on the asyncio/tulip package, but if you're ok with just bumping the requirement to 3.4 that's good with me :)

attwad commented:
That fixes #5, thanks again
